File: benchmark/src/main/java/de/hhu/bsinfo/observatory/benchmark/Operation.java
package de.hhu.bsinfo.observatory.benchmark;

import de.hhu.bsinfo.observatory.benchmark.Benchmark.Mode;
import de.hhu.bsinfo.observatory.benchmark.result.Measurement;
import de.hhu.bsinfo.observatory.benchmark.result.OverheadMeasurement;
import de.hhu.bsinfo.observatory.benchmark.result.Status;

abstract class Operation {

    private final Benchmark benchmark;
    private final Mode mode;
    private final Measurement measurement;

    private OverheadMeasurement overheadMeasurement;

    Operation(Benchmark benchmark, Mode mode, Measurement measurement) {
        this.benchmark = benchmark;
        this.mode = mode;
        this.measurement = measurement;
    }

    protected Benchmark getBenchmark() {
        return benchmark;
    }

    protected Measurement getMeasurement() {
        return measurement;
    }

    protected Mode getMode() {
        return mode;
    }

    protected OverheadMeasurement getOverheadMeasurement() {
        return overheadMeasurement;
    }

    void setOverheadMeasurement(OverheadMeasurement overheadMeasurement) {
        this.overheadMeasurement = overheadMeasurement;
    }

    abstract boolean needsFilledReceiveQueue();

    abstract Status warmUp(int operationCount);

    abstract Status execute();
}

File: cpp/src/observatory/operation/Operation.cpp
#include "Operation.h"

#include <utility>

namespace Observatory {

Operation::Operation(Observatory::Benchmark *benchmark, Observatory::Benchmark::Mode mode, std::shared_ptr<Measurement> measurement) :
        benchmark(benchmark),
        mode(mode),
        measurement(std::move(measurement)) {}

Benchmark& Operation::getBenchmark() const {
    return *benchmark;
}

Benchmark::Mode Operation::getMode() const {
    return mode;
}

OverheadMeasurement& Operation::getOverheadMeasurement() const {
    return *overheadMeasurement;
}

Measurement& Operation::getMeasurement() const {
    return *measurement;
}

void Operation::setOverheadMeasurement(std::shared_ptr<OverheadMeasurement> overheadMeasurement) {
    this->overheadMeasurement = std::move(overheadMeasurement);
}

}